pkg/virt-controller/watch/application.go

```diff
@@ -105,14 +126,64 @@ func (vca *VirtControllerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
 	vca.informerFactory.Start(stop)
-	go vca.vmController.Run(3, stop)
-	go vca.migrationController.Run(3, stop)
-	go vca.rsController.Run(3, stop)
-	httpLogger := logger.With("service", "http")
-	httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
-	if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
+	go func() {
+		httpLogger := logger.With("service", "http")
+		httpLogger.Info().Log("action", "listening", "interface", vca.host, "port", vca.port)
+		if err := http.ListenAndServe(vca.host+":"+strconv.Itoa(vca.port), nil); err != nil {
+			golog.Fatal(err)
+		}
+	}()
+
+	recorder := vca.getNewRecorder(k8sv1.NamespaceAll, leaderelectionconfig.DefaultEndpointName)
+
+	id, err := os.Hostname()
+	if err != nil {
+		golog.Fatalf("unable to get hostname: %v", err)
+	}
+
+	rl, err := resourcelock.New(vca.LeaderElection.ResourceLock,
+		leaderelectionconfig.DefaultNamespace,
+		leaderelectionconfig.DefaultEndpointName,
+		vca.clientSet.CoreV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      id,
+			EventRecorder: recorder,
+		})
+	if err != nil {
+		golog.Fatal(err)
+	}
+
+	leaderElector, err := leaderelection.NewLeaderElector(
+		leaderelection.LeaderElectionConfig{
+			Lock:          rl,
+			LeaseDuration: vca.LeaderElection.LeaseDuration.Duration,
+			RenewDeadline: vca.LeaderElection.RenewDeadline.Duration,
+			RetryPeriod:   vca.LeaderElection.RetryPeriod.Duration,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(stopCh <-chan struct{}) {
+					vca.isLeader = true
```

Comment:
I don't think a write to a boolean is an atomic operation.

Since isLeader is being accessed by multiple goroutines, it needs either to use the https://golang.org/pkg/sync/atomic/ package, or synchronize the isLeader using a lock.

Reply:
The atomic write is not the problem, since we don't need atomic compare+swap, but go is allowed to cache variables in go-routines (no volatile support). Did not think about that, when I advised the usage of the bool. 

How about using a channel and do a non-blocking select on every ready function invocation?

```golang
	case _, notReady := <- app.readyChan:
           if !notReady {
            // send 200, true
           } else {
            // send 503, false
           }

	default:
            // nothing in the channel, we are not the leader
           // send 503, false
	}
```
You would create a channel like this:

```golang
app.readyChan = make(chan struct{}, 1)
```

and when you get elected, you would simply do 

```golang
close(app.readyChan)
```
Might make sense to hide that inside a nice helper struct, with nice methods like `app.readyness.MarkReady()` and `app.readyness.IsReady()`.

Reply:
or you simply use the lock, like @davidvossel suggested.

Reply:
@rmohr @davidvossel Can you guys explain me a little bit more? Because the single place where `isLeader` written is under callback methods, and readiness method only reads it,   so I do not see here some problem. I will appreciate some example as well.

Reply:
Two goroutines are accessing a shared boolean. One goroutine is reading the value, one is writing to the value.

If the read and writes are atomic operations, this would be okay. In the case of a golang boolean, we are not guaranteed reading/writing involves a single operation. This means you could have a read collide in the middle of a write. That's why packages like this exist, https://github.com/tevino/abool, to ensure getting/setting a boolean results in a single operation. I don't know what the result of such a collision would look like, all I know is this isn't a safe way to coordinate between threads.

If you wrap the boolean access logic in a lock, everything will be fine. In general, we shouldn't be coordinating between goroutines using shared data like this. The golang way is to use a channel. I agree that a channel seems a little complex for this scenario though.

Reply:
@davidvossel thanks for the answer, I was sure that primitive types do not have such problems